llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp

```diff
@@ -861,8 +862,13 @@ void AMDGPUTargetMachine::registerPassBuilderCallbacks(PassBuilder &PB) {
             EnablePromoteKernelArguments)
           FPM.addPass(AMDGPUPromoteKernelArgumentsPass());
 
+        // Add SROA after inlining but before infer address spaces pass to
+        // unlock address space inference for smart pointers (pointers
+        // encapsulated in structs).
+        FPM.addPass(SROAPass(SROAOptions::PreserveCFG));
+
         // Add infer address spaces pass to the opt pipeline after inlining
-        // but before SROA to increase SROA opportunities.
+        // but before another SROA round to increase SROA opportunities.
```

Comment:
I'm not sure why we're running this here in the CGSCCOptimizerLate, this is more of a FunctionSimplificationPipeline sort of case

Reply:
I think this runs after inlining. In our case, we see the need of SROA to eliminate load/store of pointer-value, to enable InferAddressSpace on those pointers.

Reply:
Yes, but that's not unique among optimizations. Most optimizations should run after inlining